verilog/tools/kythe/scope_resolver.h

```diff
@@ -0,0 +1,152 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_TOOLS_KYTHE_SCOPE_RESOLVER_H_
+#define VERIBLE_VERILOG_TOOLS_KYTHE_SCOPE_RESOLVER_H_
+
+#include <map>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+#include "common/util/auto_pop_stack.h"
+#include "common/util/iterator_range.h"
+#include "verilog/tools/kythe/kythe_facts.h"
+
+namespace verilog {
+namespace kythe {
+
+// Used to wrap whatever needs to be recored in a scope item.
+struct ScopeMemberItem {
+  ScopeMemberItem(const VName& vname) : vname(vname) {}
+
+  // VNames of the members inside this scope.
+  VName vname;
+};
+
+// Represents some scope in SystemVerilog code like class or module.
+class Scope {
+ public:
+  Scope() {}
+  Scope(const Signature& signature) : signature(signature) {}
+
+  // Appedns the given vname to the members of this scope.
+  void AddMemberItem(const ScopeMemberItem& vname);
+
+  // Appends the member of the given scope to the current scope.
+  void AppendScope(const Scope& scope);
+
+  const std::vector<ScopeMemberItem>& Members() const { return members; }
+  const Signature& GetSignature() const { return signature; }
+
+  // Searches for the given reference_name in the current scope and returns its
+  // VName.
+  const VName* SearchForDefinition(absl::string_view name) const;
```

Comment:
Can this be changed to VName&? Maybe return Optional<VName> or StatusOr<VName> if it's nullable (it's a largish change, no need to do it now).